core/server/server.js

```diff
@@ -414,14 +432,25 @@ class Server {
     log(`Server is starting up: ${this.baseUrl}`)
 
     const camp = (this.camp = Camp.create({
-      documentRoot: path.resolve(__dirname, '..', '..', 'public'),
+      documentRoot: this.config.public.documentRoot,
       port,
       hostname,
       secure,
       cert,
       key,
     }))
 
+    const index = getHandlerIndex(this.camp, 'staticLayer')
+    addHandlerAtIndex(this.camp, index, function addCacheHeaders(
```

Comment:
This will run for everything that isn't handled by a route, which I think mainly is the `camp.notfound` route defined in `server.js`.

An alternative could be to replace `staticLayer` with our own function, although I don't see a way to do that without copying in a bunch of code from `respondWithFile`.

Hmm, Scoutcamp is sending some cache headers on files…

https://github.com/espadrine/sc/blob/c1042e58fb849e1a2ec87dc22695d21329b0e8e1/lib/camp.js#L677-L685

Reply:
Yes - inserting the handler here will apply those headers to every subsequent handler in the stack (i.e: also `notFound`). Is there a problem if we serve 404s with these headers applied?

> Any thoughts on this part?

I've not tried it yet. IIUC, doing that would also require us to _replace_ `staticLayer`, yeah?

Reply:
It makes me wonder if the headers Scoutcamp is setting (in that link) are the ones we want.

Reply:
..so you're thinking what we actually want to do is just bin all this off an go with the out-of-the-box ScoutCamp behaviour?

Reply:
This is from the output of `curl -v https://shields-staging.herokuapp.com/ > /dev/null`:

```
> GET / HTTP/1.1
> Host: shields-staging.herokuapp.com
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: Cowboy
< Connection: keep-alive
< Content-Type: text/html
< Last-Modified: Thu, 01 Oct 2020 02:21:59 GMT
< Date: Thu, 01 Oct 2020 19:21:10 GMT
< Transfer-Encoding: chunked
< Via: 1.1 vegur
```

Comparing that to our own static content (`curl -v https://shields-staging.herokuapp.com/badge/foo-bar-blue > /dev/null` …

```
> GET /badge/foo-bar-blue HTTP/1.1
> Host: shields-staging.herokuapp.com
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< Server: Cowboy
< Connection: keep-alive
< Cache-Control: max-age=86400, s-maxage=86400
< Last-Modified: Thu, 01 Oct 2020 02:24:01 GMT
< Content-Type: image/svg+xml;charset=utf-8
< Date: Thu, 01 Oct 2020 19:23:08 GMT
< Transfer-Encoding: chunked
< Via: 1.1 vegur
```

… the thing that's missing is an actual Cache-Control header. I think by default that means it's set to `private`, and that we'd do better by additionally setting `max-age`.

The changes you've made here for testing seem great.

I get not wanting to modify the work, although I still don't see a way to keep the code you're adding from running on the requests that are further down the stack, like `camp.notfound`. The support for If-Modified-Since + 304's in sc seems good. What would you think about adding a staticCacheSeconds param to Scoutcamp? Or even hard-coding it, since it's a just-for-us fork of something we basically want to deprecate anyway.

Reply:
> and that we'd do better by additionally setting `max-age`

OK.. so then, isn't that what this PR does at the moment? The current default behaviour plus setting a max-age (and s-max-age and expires) Are we having two different conversations?

I've said

> Yes - inserting the handler here will apply those headers to every subsequent handler in the stack (i.e: also `notFound`). Is there a problem if we serve 404s with these headers applied?

we've said a bunch more stuff that doesn't really answer that question (but feels like we've circled back round to "what this PR should do is what it does now", or I've misunderstood?) and now we're back to

> although I still don't see a way to keep the code you're adding from running on the requests that are further down the stack, like `camp.notfound`

Can you really clearly say what you're asking/saying/suggesting on this topic. There is something important missing. Fortunately cache invalidation is a notoriously easy topic, so it should be easy to sort out ;)

Reply:
> Fortunately cache invalidation is a notoriously easy topic, so it should be easy to sort out ;)

🤣 

> Can you really clearly say what you're asking/saying/suggesting on this topic. There is something important missing. 

Sorry I've haven't been explaining it clearly.

The code you've added here causes `setHeadersForCacheLength()` to be invoked on requests for static content, as intended, and for requests that don't match static content which are handled by `camp.notfound()`, which is unintended.

We should ensure we're only adding headers to the requests fort static content.

The only correct way I see to do this is to modify scoutcamp's `replyWithFile()` function with additional code to set max-age headers. The clean way to configure it would be to pass a staticCacheSeconds option to the Camp constructor, set it as an instance variable (something like `server.staticCacheSeconds`), and then pass it as a parameter to `respondWithFile()` on https://github.com/badges/sc/blob/35f6b68ae890fd0c8f3771c74c9d7971b69cee41/lib/camp.js#L701.

Reply:
> The code you've added here causes `setHeadersForCacheLength()` to be invoked on requests for static content, as intended, and for requests that don't match static content which are handled by `camp.notfound()`

Exactly. This is what mean when I say:

> Yes - inserting the handler here will apply those headers to every subsequent handler in the stack (i.e: also `notFound`)

This is going round in the same circle again: I'm saying "yes - the max-age header will also be set on 404 responses" and then you're replying with "the max-age header will be set on 404 responses". :laughing: 

The question I'm asking is: If we serve a 404 with the same (fairly short) `max-age` as the rest of the front-end, is that actually a problem? i.e: What is the bad thing that happens if we do that? As far as I can tell, its not inherently wrong to set a `max-age` on a 404, but I'd like to understand what I'm missing about why its going to cause us a problem in this situation. I'm not saying it necessarily _isn't_ - I'm just not seeing what you're seeing.

Reply:
> This is going round in the same circle again: I'm saying "yes - the max-age header will also be set on 404 responses" and then you're replying with "the max-age header will be set on 404 responses". 😆

😆 

> The question I'm asking is: If we serve a 404 with the same (fairly short) `max-age` as the rest of the front-end, is that actually a problem? i.e: What is the bad thing that happens if we do that? As far as I can tell, its not inherently wrong to set a `max-age` on a 404, but I'd like to understand what I'm missing about why its going to cause us a problem in this situation. I'm not saying it necessarily _isn't_ - I'm just not seeing what you're seeing.

What I can say now is: setting cache headers on 404 responses _is not the intention of this PR_, but would be an unintended effect.

I can investigate this effect further, and articulate whether or not there's a circumstance where it will cause a problem. (I'm mostly thinking about invalid badge URLs that are 404's.)

Reply:
I find this use of the word "unintended" annoying because it seems like you're still trying to say I don't know what the code I've written does, but I'm going to take the most charitable interpretation of this and assume you mean you're saying it doesn't do what you want, rather than I don't know what it does.

Either way, I care about getting to a stage where we can do one-step deploys more than I care what TTL we're serving the 404 badge with. Lets move this on and say we don't want to make that behaviour change at this time. I've submitted https://github.com/badges/sc/pull/3 and https://github.com/badges/sc/pull/4 . I'm going to close this PR and I'll submit a revised PR once we've made a change upstream.

We can defer the issue of whether we do or don't want to send an explicit `max-age` on 404s. At the CloudFlare level I think the behaviour we _should_ have if we send no `max-age` is that the 404 badge gets implicitly cached by CloudFlare with a 3m TTL (that's my reading of https://support.cloudflare.com/hc/en-us/articles/200172516-Understanding-Cloudflare-s-CDN#h_51422705-42d0-450d-8eb1-5321dcadb5bc anyway) so I don't think it should actually be a cache miss on every request anyway (at least at the CDN level).. but then I went to check that and found they are served with `cf-cache-status: DYNAMIC`. Then I started pulling the thread and realised there's loads of stuff not getting served from cache that should be so I guess they _are_ always a cache miss, but it seems like what TTL we've got on the 404s might not be the biggest issue in the grand scheme of things. Anyway, lets kick this 404 can down the road and leave them with no explicit `max-age` for now..